README.txt
/*******************************************
*
*	UPlayer version 1.0
*
*   Author: Adam Oleksy
*   date: 16.03.2011
*
********************************************/


Description:
//////////////////////////////////////////
Module to play wav files. It uses SDL library.

Install:
//////////////////////////////////////////
1. Download SDL library from http://www.libsdl.org/ and
install them in the system (ex. C:\Program Files\SDL-1.2.14).

2. Under Windows platform add to system Enviroment Variables:
 - to PATH add sdl directory  (ex. C:\Program Files\SDL-1.2.14)

3. For compile sources you have to download and install CMake also. 
CMake website is http://www.cmake.org  
a. Set system Enviroment Variables:
 - SDLDIR (ex. C:\Program Files\SDL-1.2.14);
 - URBI_ROOT (ex. C:\Program Files\Gostai Engine Runtime\2.7\).
b. use CMake to generate and configure the project.


Module functions:
//////////////////////////////////////////
UPlayer.play("sound.wav"); - play the sound, you can run this
function in background;
UPlayer.stop; - stop the sound;
UPlayer.isPlaying; - returns true (1) if sound is playing.

Urbiscript example:
//////////////////////////////////////////

loadModule("UPlayer");
var player=UPlayer.new;
player.play("music.wav"),





debian/changelog
liburbisound2.7 (0.0.1-0ubuntu0ppa3~lucid1) lucid; urgency=low

  * Poprawienie katalogu

 -- Adam Oleksy <adam.oleksy@pwr.wroc.pl>  Tue, 22 Nov 2011 17:06:57 +0100

liburbisound2.7 (0.0.1-0ubuntu0ppa2~lucid1) lucid; urgency=low

  * Biblioteka instaluje sie w /usr/lib/gostai/uobjects

 -- Adam Oleksy <adam.oleksy@pwr.wroc.pl>  Tue, 22 Nov 2011 14:15:32 +0100

liburbisound2.7 (0.0.1-0ubuntu0ppa1~lucid1) lucid; urgency=low

  * Initial release

 -- Adam Oleksy <adam.oleksy@pwr.wroc.pl>  Wed, 27 Apr 2011 14:49:20 +0200


src/sdlwave.c
/* 
 * File:   sdlwave.c
 * Author: aoleksy
 *
 * Created on 6 marzec 2011, 20:31
 */

#include "SDL.h"
#include "SDL_audio.h"

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>

/*
 * 
 */
void PlaySound(char *file);

int main(int argc, char** argv) {
    extern void mixaudio(void *unused, Uint8 *stream, int len);
    
    SDL_AudioSpec fmt;

    /* Set 16-bit stereo audio at 22Khz */
    fmt.freq = 22050;
    fmt.format = AUDIO_S16;
    fmt.channels = 2;
    fmt.samples = 512;        /* A good value for games */
    fmt.callback = mixaudio;
    fmt.userdata = NULL;

    /* Open the audio device and start playing sound! */
    if ( SDL_OpenAudio(&fmt, NULL) < 0 ) {
        fprintf(stderr, "Unable to open audio: %s\n", SDL_GetError());
        exit(1);
    }
    SDL_PauseAudio(0);
    //PlaySound("/tmp/test.wav");
    SDL_Delay(1000);
    SDL_CloseAudio();
    return (EXIT_SUCCESS);
}

#define NUM_SOUNDS 2
struct sample {
    Uint8 *data;
    Uint32 dpos;
    Uint32 dlen;
} sounds[NUM_SOUNDS];

void mixaudio(void *unused, Uint8 *stream, int len)
{
    int i;
    Uint32 amount;
    for ( i=0; i<NUM_SOUNDS; ++i ) {
        printf("aaa %d\n", i);
        amount = (sounds[i].dlen-sounds[i].dpos);
        if ( amount > len ) {
            amount = len;
        }
        SDL_MixAudio(stream, &sounds[i].data[sounds[i].dpos], amount, SDL_MIX_MAXVOLUME);
        sounds[i].dpos += amount;
    }
}

void PlaySound(char *file)
{
    int index;
    SDL_AudioSpec wave;
    Uint8 *data;
    Uint32 dlen;
    SDL_AudioCVT cvt;

    /* Look for an empty (or finished) sound slot */
    for ( index=0; index<NUM_SOUNDS; ++index ) {
        if ( sounds[index].dpos == sounds[index].dlen ) {
            break;
        }
    }
    if ( index == NUM_SOUNDS )
        return;

    /* Load the sound file and convert it to 16-bit stereo at 22kHz */
    if ( SDL_LoadWAV(file, &wave, &data, &dlen) == NULL ) {
        fprintf(stderr, "Couldn't load %s: %s\n", file, SDL_GetError());
        return;
    }
    SDL_BuildAudioCVT(&cvt, wave.format, wave.channels, wave.freq,
                            AUDIO_S16,   2,             22050);
    cvt.buf = malloc(dlen*cvt.len_mult);
    memcpy(cvt.buf, data, dlen);
    cvt.len = dlen;
    SDL_ConvertAudio(&cvt);
    SDL_FreeWAV(data);

    /* Put the sound data in the slot (it starts playing immediately) */
    if ( sounds[index].data ) {
        free(sounds[index].data);
    }
    SDL_LockAudio();
    sounds[index].data = cvt.buf;
    sounds[index].dlen = cvt.len_cvt;
    sounds[index].dpos = 0;
    SDL_UnlockAudio();
}

src/UrbiSound.cpp
/* 
 * File:   UrbiSound.cpp
 * Author: aoleksy
 * 
 * Created on 7 marzec 2011, 11:02
 */

#include "UrbiSound.h"

#include "SDL.h"
#include "SDL_audio.h"

#include <boost/shared_ptr.hpp>
#include <boost/noncopyable.hpp>

#include <iostream>
#include <vector>
#include <list>

using namespace urbi;
using namespace std;
using namespace boost;

void mixAudio(void *, Uint8 *stream, int len);

typedef shared_ptr<Uint8> Uint8Pointer;

class Sample {
private:
    Uint8Pointer mData;
    Uint32 mDataLen;
    Uint32 mDataPos;
    const UPlayer* mOwner;
public:
    Sample(const UPlayer* owner, Uint8Pointer data, Uint32 dataLen)
    : mOwner(owner), mData(data), mDataLen(dataLen), mDataPos(0) {
	}
    
    const UPlayer* getOwner() const {
        return mOwner;
    }
    
    const Uint8* getData(int howMany) {
        Uint8* ret = &mData.get()[mDataPos];
        mDataPos += howMany;
        return ret;
    }
    
    int getDataLeft() const {
        return mDataLen - mDataPos;
    }
};
class SampleBelongsTo {
private:
    const UPlayer* mUrbiSound;
public:
    SampleBelongsTo(const UPlayer* urbiSound) : mUrbiSound(urbiSound) { }
    bool operator() (const Sample& sample) {
        return mUrbiSound == sample.getOwner();
    }
};

typedef list<Sample> SampleList;

class SDLSoundSingleton : private boost::noncopyable {
private:
    SDLSoundSingleton();
    
    SDL_AudioSpec mFMT;
    bool mIsOpened;
    
    SampleList mSampleList;
    
public:
    ~SDLSoundSingleton();
    
    static SDLSoundSingleton& getInstance();
   
    bool openDevice();
    void closeDevice();
    bool isOpened();
    bool stopSample(const UPlayer* owner);
    
    SampleList &getSampleList();
    void addSample(const Sample& sample);
    bool play(const UPlayer* owner, const string& file);
    void stop(const UPlayer* owner);
};

// Implementations
void mixAudio(void *, Uint8 *stream, int len) {
    SampleList& sampleList = SDLSoundSingleton::getInstance().getSampleList();
    SampleList::iterator i = sampleList.begin();
    while(i != sampleList.end()) {
        int amount = static_cast<int>(i->getDataLeft());
        amount = amount < len ? amount : len;
        SDL_MixAudio(stream, i->getData(amount), amount, SDL_MIX_MAXVOLUME);
        
        //Iteracja pętli oraz sprawdzenie czy w sample cos jeszcze zostalo
        SampleList::iterator previous = i++;
        if(!previous->getDataLeft())
            sampleList.erase(previous);
    }
    
    if(sampleList.empty())
        SDL_PauseAudio(1);
}

//------------------------- SDLSoundSingleton ----------------------------//

SDLSoundSingleton::SDLSoundSingleton() : mIsOpened(false) {

    /* Set 16-bit stereo audio at 22Khz */
    mFMT.freq = 44100;
    mFMT.format = AUDIO_S16;
    mFMT.channels = 2;
    mFMT.samples = 512; /* A good value for games */
    mFMT.callback = mixAudio;
    mFMT.userdata = NULL;
}

SDLSoundSingleton::~SDLSoundSingleton() {
    SDL_CloseAudio();
}

SDLSoundSingleton& SDLSoundSingleton::getInstance() {
    static SDLSoundSingleton instance;
    return instance;
}

inline bool SDLSoundSingleton::openDevice() {
    if (!isOpened()) {
        if(SDL_OpenAudio(&mFMT, NULL) < 0) {
            cerr << "Could'nt open audio: " <<SDL_GetError() << endl;
            return false;
        }
    }
        
    return (mIsOpened = true);
}

inline void SDLSoundSingleton::closeDevice() {
    if (isOpened()) {
        mIsOpened = false;
        SDL_CloseAudio();
    }
}

inline bool SDLSoundSingleton::isOpened() {
    return mIsOpened;
}

bool SDLSoundSingleton::play(const UPlayer* owner, const string& file) {
    if(!openDevice())
        return false;
    
    SDL_AudioSpec wave;
    Uint8 *data;
    Uint32 dlen;
    if (SDL_LoadWAV(file.c_str(), &wave, &data, &dlen) == NULL) {
        cerr << "Couldn't load " << file << ": " << SDL_GetError() << endl;
        return false;
    }

    SDL_AudioCVT cvt;
    SDL_BuildAudioCVT(&cvt, wave.format, wave.channels, wave.freq,
            AUDIO_S16, 2, 44100);
    cvt.buf = new Uint8[dlen * cvt.len_mult];
    memcpy(cvt.buf, data, dlen);
    cvt.len = dlen;
    SDL_ConvertAudio(&cvt);
    SDL_FreeWAV(data);
    
    stop(owner);
    SDL_LockAudio();
    Uint8Pointer sampleData(cvt.buf);
    mSampleList.push_back(Sample(owner, sampleData, cvt.len_cvt));
    SDL_PauseAudio(0);
    SDL_UnlockAudio();
    
    return true;
}

inline void SDLSoundSingleton::stop(const UPlayer* owner) {
    SDL_LockAudio();
    mSampleList.remove_if(SampleBelongsTo(owner));
    SDL_UnlockAudio();
}

inline SampleList &SDLSoundSingleton::getSampleList() {
    return mSampleList;
}


//------------------------- UrbiSound ----------------------------//

UPlayer::UPlayer(const std::string& name) : UObject(name) {
    //Bind the functions
    UBindFunction(UPlayer, play);
    UBindFunction(UPlayer, stop);
    UBindFunction(UPlayer, isPlaying);
}

UPlayer::~UPlayer() {
}

bool UPlayer::play(const std::string& file) {
    return SDLSoundSingleton::getInstance().play(this, file);
}

void UPlayer::stop() {
    SDLSoundSingleton::getInstance().stop(this);
    return;
}

bool UPlayer::isPlaying() {
    return true;
}

src/UPlayer.cpp
/* 
 * File:   UrbiSound.cpp
 * Author: aoleksy
 * 
 * Created on 7 marzec 2011, 11:02
 */

#include "UPlayer.h"

#include "SDL.h"
#include "SDL_audio.h"

#include <boost/shared_ptr.hpp>
#include <boost/noncopyable.hpp>

#include <iostream>
#include <vector>
#include <list>
#include <algorithm>

using namespace urbi;
using namespace std;
using namespace boost;

void mixAudio(void *, Uint8 *stream, int len);

typedef shared_ptr<Uint8> Uint8Pointer;

class Sample {
private:
    Uint8Pointer mData;
    Uint32 mDataLen;
    Uint32 mDataPos;
    const UPlayer* mOwner;
public:
    Sample(const UPlayer* owner, Uint8Pointer data, Uint32 dataLen)
    : mOwner(owner), mData(data), mDataLen(dataLen), mDataPos(0) {
        mOwner->lockPlay(true);
    }
    
    ~Sample() {
        mOwner->lockPlay(false);
    }
    
    const UPlayer* getOwner() const {
        return mOwner;
    }
    
    const Uint8* getData(int howMany) {
        Uint8* ret = &mData.get()[mDataPos];
        mDataPos += howMany;
        return ret;
    }
    
    int getDataLeft() const {
        return mDataLen - mDataPos;
    }
};

typedef shared_ptr<Sample> SamplePointer;

class SampleBelongsTo {
private:
    const UPlayer* mUrbiSound;
public:
    SampleBelongsTo(const UPlayer* urbiSound) : mUrbiSound(urbiSound) { }
    bool operator() (const SamplePointer& sample) {
        return mUrbiSound == sample->getOwner();
    }
};

typedef list<SamplePointer> SampleList;

class SDLSoundSingleton : private boost::noncopyable {
private:
    SDLSoundSingleton();
    
    SDL_AudioSpec mFMT;
    bool mIsOpened;
    
    SampleList mSampleList;
    
public:
    ~SDLSoundSingleton();
    
    static SDLSoundSingleton& getInstance();
   
    bool openDevice();
    void closeDevice();
    bool isOpened();
    bool stopSample(const UPlayer* owner);
    
    SampleList &getSampleList();
    void addSample(const Sample& sample);
    bool play(const UPlayer* owner, const string& file);
    void stop(const UPlayer* owner);
    bool isPlaying(const UPlayer* owner);
};

// Implementations
void mixAudio(void *, Uint8 *stream, int len) {
    SampleList& sampleList = SDLSoundSingleton::getInstance().getSampleList();
    SampleList::iterator i = sampleList.begin();
    while(i != sampleList.end()) {
        int amount = static_cast<int>((*i)->getDataLeft());
        amount = amount < len ? amount : len;
        SDL_MixAudio(stream, (*i)->getData(amount), amount, SDL_MIX_MAXVOLUME);
        
        //Iteracja pętli oraz sprawdzenie czy w sample cos jeszcze zostalo
        SampleList::iterator previous = i++;
        if(!(*previous)->getDataLeft())
            sampleList.erase(previous);
    }
    
    if(sampleList.empty())
        SDL_PauseAudio(1);
}

//------------------------- SDLSoundSingleton ----------------------------//

SDLSoundSingleton::SDLSoundSingleton() : mIsOpened(false) {

    /* Set 16-bit stereo audio at 22Khz */
    mFMT.freq = 44100;
    mFMT.format = AUDIO_S16;
    mFMT.channels = 2;
    mFMT.samples = 512; /* A good value for games */
    mFMT.callback = mixAudio;
    mFMT.userdata = NULL;
}

SDLSoundSingleton::~SDLSoundSingleton() {
    SDL_CloseAudio();
}

SDLSoundSingleton& SDLSoundSingleton::getInstance() {
    static SDLSoundSingleton instance;
    return instance;
}

inline bool SDLSoundSingleton::openDevice() {
    if (!isOpened()) {
        if(SDL_OpenAudio(&mFMT, NULL) < 0) {
            cerr << "Could'nt open audio: " <<SDL_GetError() << endl;
            return false;
        }
    }
        
    return (mIsOpened = true);
}

inline void SDLSoundSingleton::closeDevice() {
    if (isOpened()) {
        mIsOpened = false;
        SDL_CloseAudio();
    }
}

inline bool SDLSoundSingleton::isOpened() {
    return mIsOpened;
}

bool SDLSoundSingleton::play(const UPlayer* owner, const string& file) {
    if(!openDevice())
        return false;
    
    SDL_AudioSpec wave;
    Uint8 *data;
    Uint32 dlen;
    if (SDL_LoadWAV(file.c_str(), &wave, &data, &dlen) == NULL) {
        cerr << "Couldn't load " << file << ": " << SDL_GetError() << endl;
        return false;
    }

    SDL_AudioCVT cvt;
    SDL_BuildAudioCVT(&cvt, wave.format, wave.channels, wave.freq,
            AUDIO_S16, 2, 44100);
    cvt.buf = new Uint8[dlen * cvt.len_mult];
    memcpy(cvt.buf, data, dlen);
    cvt.len = dlen;
    SDL_ConvertAudio(&cvt);
    SDL_FreeWAV(data);
    
    stop(owner);
    SDL_LockAudio();
    Uint8Pointer sampleData(cvt.buf);
    SamplePointer samplePointer(new Sample(owner, sampleData, cvt.len_cvt));
    mSampleList.push_back(samplePointer);
    SDL_PauseAudio(0);
    SDL_UnlockAudio();
    
    return true;
}

inline void SDLSoundSingleton::stop(const UPlayer* owner) {
    SDL_LockAudio();
    mSampleList.remove_if(SampleBelongsTo(owner));
    SDL_UnlockAudio();
}

bool SDLSoundSingleton::isPlaying(const UPlayer* owner) {
    return find_if(mSampleList.begin(),
            mSampleList.end(),
            SampleBelongsTo(owner)) != mSampleList.end();
}

inline SampleList &SDLSoundSingleton::getSampleList() {
    return mSampleList;
}


//------------------------- UrbiSound ----------------------------//

UPlayer::UPlayer(const std::string& name) : UObject(name) {
    //Bind the functions
    UBindThreadedFunction(UPlayer, play, LOCK_NONE);
    UBindFunction(UPlayer, stop);
    UBindFunction(UPlayer, isPlaying);
}

UPlayer::~UPlayer() {
}

bool UPlayer::play(const std::string& file) {
    bool ret = SDLSoundSingleton::getInstance().play(this, file);
    mLockPlay.lock();
    mLockPlay.unlock();
    return ret;
}

void UPlayer::stop() {
    SDLSoundSingleton::getInstance().stop(this);
    return;
}

bool UPlayer::isPlaying() {
    return SDLSoundSingleton::getInstance().isPlaying(this);
}

void UPlayer::lockPlay(bool lock) const {
    if (lock)
        mLockPlay.lock();
    else
        mLockPlay.unlock();
}